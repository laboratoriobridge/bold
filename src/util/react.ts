import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
