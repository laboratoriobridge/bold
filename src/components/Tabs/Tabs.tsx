import React, { CSSProperties } from 'react'

import { useRovingTabIndex } from '../../hooks/useRovingTabIndex'
import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'

export interface TabsProps extends Omit<React.HTMLAttributes<HTMLUListElement>, 'style'> {
  style?: ExternalStyles
}

export function Tabs(props: TabsProps) {
  const { style, ...rest } = props
  const { classes, css } = useStyles(createStyles)

  const rootRef = useRovingTabIndex({
    getItems: root => Array.from(root.querySelectorAll(`[role="tab"]:not(.disabled)`)),
  })

  return <ul ref={rootRef} className={css(classes.ul, style)} role='tablist' {...rest} />
}

const createStyles = (theme: Theme) => ({
  ul: {
    listStyle: 'none',
    padding: '0',
    margin: '0',
    fontSize: theme.typography.sizes.button,
  } as CSSProperties,
})
