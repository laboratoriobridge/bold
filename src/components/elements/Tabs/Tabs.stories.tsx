import { storiesOf } from '@storybook/react'
import React from 'react'

import { withRouter } from '../../../stories-addons/withRouter'

import { TabLink, Tabs } from './Tabs'

storiesOf('Components|Tabs', module)
  // @ts-ignore
  .addParameters({
    info: {
      text: `
                Active é obtido através da prop <code>active</code> ou caso ocorra um match entre a URL
                atual e o link (Quarto item).
            `,
    },
  })
  .addDecorator(withRouter(['/test-4'], 1))
  .add('default', () => (
    <Tabs>
      <TabLink to='/test-1'>First item</TabLink>
      <TabLink to='/test-2' active>
        Second item
      </TabLink>
      <TabLink to='/test-3'>Third item</TabLink>
      <TabLink to='/test-4'>Fourth item</TabLink>
      <TabLink to='/test-5' disabled>
        Disabled item
      </TabLink>
    </Tabs>
  ))
