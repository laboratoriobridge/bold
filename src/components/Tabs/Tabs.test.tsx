import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { TabItem } from './TabItem'
import { Tabs } from './Tabs'

it('renders correctly', () => {
  expect(
    render(
      <Tabs>
        <TabItem>Home</TabItem>
        <TabItem>Test</TabItem>
        <TabItem active>Active</TabItem>
        <TabItem disabled>Disabled</TabItem>
      </Tabs>
    ).container
  ).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  expect(render(<Tabs style={{ color: 'red' }}>Children</Tabs>).container).toMatchSnapshot()
})

it('should have a roving tabindex', () => {
  const { container } = render(
    <Tabs>
      <TabItem>Home</TabItem>
      <TabItem>Test</TabItem>
      <TabItem active>Active</TabItem>
      <TabItem disabled>Disabled</TabItem>
    </Tabs>
  )
  const ul = container.querySelector('ul')
  const items = container.querySelectorAll<any>('[role="tab"]')
  expect(items[0].getAttribute('tabindex')).toEqual('0')
  expect(items[1].getAttribute('tabindex')).toEqual('-1')

  items[2].focus()
  expect(items[0].getAttribute('tabindex')).toEqual('-1')
  expect(items[1].getAttribute('tabindex')).toEqual('-1')
  expect(items[2].getAttribute('tabindex')).toEqual('0')

  fireEvent.keyDown(ul, { key: 'ArrowDown' })
  expect(items[2].getAttribute('tabindex')).toEqual('0')

  fireEvent.keyDown(ul, { key: 'ArrowUp' })
  expect(items[1].getAttribute('tabindex')).toEqual('0')
  expect(items[2].getAttribute('tabindex')).toEqual('-1')
})
