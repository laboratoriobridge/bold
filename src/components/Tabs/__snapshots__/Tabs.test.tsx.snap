// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.emotion-8 {
  list-style: none;
  padding: 0;
  margin: 0;
  font-size: 0.875rem;
}

.emotion-1 {
  display: inline-block;
}

.emotion-0 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #3A3A47;
  font-weight: bold;
  padding: 0.5rem 0.75rem;
  line-height: 1rem;
  border-radius: 4px;
  outline: 0;
  border: 2px solid transparent;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
}

.emotion-0:not(.disabled):hover {
  cursor: pointer;
  background: #F0F0F5;
}

.emotion-0:not(.disabled):active {
  background: #F0F0F5;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-0:focus {
  border-color: #0069D0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

.emotion-4 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #3A3A47;
  font-weight: bold;
  padding: 0.5rem 0.75rem;
  line-height: 1rem;
  border-radius: 4px;
  outline: 0;
  border: 2px solid transparent;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
  color: #0069D0;
  border-bottom-color: #0069D0;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-4:not(.disabled):hover {
  cursor: pointer;
  background: #F0F0F5;
}

.emotion-4:not(.disabled):active {
  background: #F0F0F5;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-4:focus {
  border-color: #0069D0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

.emotion-6 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #3A3A47;
  font-weight: bold;
  padding: 0.5rem 0.75rem;
  line-height: 1rem;
  border-radius: 4px;
  outline: 0;
  border: 2px solid transparent;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
  color: #AAAAB9;
}

.emotion-6:not(.disabled):hover {
  cursor: pointer;
  background: #F0F0F5;
}

.emotion-6:not(.disabled):active {
  background: #F0F0F5;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-6:focus {
  border-color: #0069D0;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

.emotion-6:hover {
  cursor: not-allowed;
}

<div>
  <ul
    class="emotion-8"
    role="tablist"
  >
    <li
      class="emotion-1"
    >
      <span
        class="emotion-0"
        role="tab"
        tabindex="0"
      >
        Home
      </span>
    </li>
    <li
      class="emotion-1"
    >
      <span
        class="emotion-0"
        role="tab"
        tabindex="-1"
      >
        Test
      </span>
    </li>
    <li
      class="emotion-1"
    >
      <span
        class="emotion-4"
        role="tab"
        tabindex="-1"
      >
        Active
      </span>
    </li>
    <li
      class="emotion-1"
    >
      <span
        class="emotion-6 disabled"
        role="tab"
      >
        Disabled
      </span>
    </li>
  </ul>
</div>
`;

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  list-style: none;
  padding: 0;
  margin: 0;
  font-size: 0.875rem;
  color: red;
}

<div>
  <ul
    class="emotion-0"
    role="tablist"
  >
    Children
  </ul>
</div>
`;
